### PR TITLE
Filtered Space Saving for top-K queries

### PR DESCRIPTION
**Pull request by usmanm**

### CODE REVIEW

**Comment by derekjn** [comment]
Can you add some .sql tests that verify errors are properly thrown? e.g. for by-ref args.

File: src/backend/pipeline/fss.c [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+	FSS *fss;
+
+	if (k > m)
+		elog(ERROR, "maximum value of k exceeded");
+
+	/* TODO(usmanm): Check bounds for k, m, h */
+
+	/* TODO(usmanm): Add support for ref types */
+	if (!typ->typbyval)
+		elog(ERROR, "fss doesn't support types by reference");
```
Comment (derekjn) [inline src/backend/pipeline/fss.c]: This should probably be forced from the .sql tests.

Comment (usmanm) [inline src/backend/pipeline/fss.c]: Yea, trying to figure out how to make this work. There's not way to tell the parser to only allow types by reference, so need some hackery.

File: src/backend/utils/adt/fssfuncs.c [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+Datum
+fss_merge_agg_trans(PG_FUNCTION_ARGS)
+{
+	MemoryContext old;
+	MemoryContext context;
+	FSS *state;
+	FSS *incoming = fss_recv(PG_GETARG_VARLENA_P(1));
+
+	if (!AggCheckCallContext(fcinfo, &context))
+		elog(ERROR, "fss_union_agg_trans called in non-aggregate context");
```
Comment (derekjn) [inline src/backend/utils/adt/fssfuncs.c]: `fss_merge_agg_trans`

File: src/backend/utils/adt/fssfuncs.c [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+#include "utils/bytea.h"
+#include "utils/builtins.h"
+#include "utils/datum.h"
+#include "utils/fssfuncs.h"
+#include "utils/typcache.h"
+
+#define DEFAULT_K 5
+
+static FSS *
+fss_recv(struct varlena *bytes)
```
Comment (derekjn) [inline src/backend/utils/adt/fssfuncs.c]: This should probably be renamed because it's not a `*recv` function in the typical sense.

**Comment by derekjn** [comment]
https://github.com/pipelinedb/docs/issues/25

**Pull request closed without merging**